File: mobile/ios/Vault/VaultManager.swift
import Foundation

/// Process-global vault session shared by the app and the Credential Provider
/// extension. Holds the unlocked native vault (UniFFI) in memory only.
@MainActor
final class VaultManager: ObservableObject {
    static let shared = VaultManager()

    enum SyncState { case synced, pending, error, offline }

    @Published private(set) var unlocked = false
    @Published private(set) var syncState: SyncState = .synced

    private var vault: VaultHandle?
    private var api: ApiClient?
    private let cache = VaultCache()
    private var cursor: Int = 0
    private var baseVersions: [String: Int] = [:]

    var hasCache: Bool { cache.hasCache }

    // MARK: unlock paths

    func unlockOnline(instanceUrl: String, username: String, password: String, totp: String?) async throws {
        let client = ApiClient(baseUrl: instanceUrl.trimmingTrailingSlash())
        let ls = try jsonObject(opaqueLoginStart(password: password))
        let start = try await client.loginStart(username: username, credentialRequest: ls["message"] as! String)
        let finalization = try opaqueLoginFinish(
            stateB64: ls["state"] as! String, password: password,
            responseB64: start["credential_response"] as! String)
        let outcome = try await client.loginFinish(
            flowId: start["flow_id"] as! String, finalization: finalization,
            deviceName: "iOS", totp: totp)
        guard let token = outcome["access_token"] as? String else { throw VaultError.Message(message: "second factor required") }
        client.accessToken = token

        let crypto = try await client.accountCrypto()
        let pulled = try await client.pull(cursor: 0)
        cursor = pulled["cursor"] as! Int
        let records = pulled["records"] as! [[String: Any]]
        rememberBaseVersions(records)
        let cryptoStr = jsonString(crypto)
        let recordsStr = jsonString(records)
        vault = try VaultHandle.unlock(password: password, cryptoJson: cryptoStr, recordsJson: recordsStr)
        api = client
        cache.save(cryptoJson: cryptoStr, recordsJson: recordsStr)
        syncState = .synced
        unlocked = true
    }

    func unlockOffline(password: String) throws {
        guard let crypto = cache.cryptoJson else { throw VaultError.Message(message: "no cache") }
        let records = cache.recordsJson ?? "[]"
        vault = try VaultHandle.unlock(password: password, cryptoJson: crypto, recordsJson: records)
        rememberBaseVersions(jsonArray(records))
        syncState = .offline
        unlocked = true
    }

    func unlockBiometric(accountKeyB64: String) throws {
        guard let crypto = cache.cryptoJson else { throw VaultError.Message(message: "no cache") }
        let records = cache.recordsJson ?? "[]"
        vault = try VaultHandle.unlockWithAccountKey(accountKeyB64: accountKeyB64, cryptoJson: crypto, recordsJson: records)
        rememberBaseVersions(jsonArray(records))
        syncState = .offline
        unlocked = true
    }

    func exportAccountKey() throws -> String { try requireVault().exportAccountKey() }

    func lock() {
        vault = nil
        api = nil
        baseVersions.removeAll()
        cursor = 0
        unlocked = false
    }

    // MARK: items

    private func requireVault() throws -> VaultHandle {
        guard let v = vault else { throw VaultError.Message(message: "locked") }
        return v
    }

    func listActive() -> [String] { (try? requireVault().listActive()) ?? [] }
    func search(_ q: String) -> [String] { (try? requireVault().search(query: q)) ?? [] }
    func candidates(for url: String) -> [String] { (try? requireVault().candidatesFor(url: url)) ?? [] }
    func item(_ id: String) -> [String: Any] {
        guard let json = try? requireVault().getItem(id: id) else { return [:] }
        return (try? jsonObject(json)) ?? [:]
    }

    func generatePassword() -> String {
        (try? Vault.generatePassword(optionsJson: #"{"length":20,"lowercase":true,"uppercase":true,"digits":true,"symbols":true}"#)) ?? ""
    }

    func sync() async {
        guard let client = api else { syncState = .offline; return }
        do {
            let pulled = try await client.pull(cursor: cursor)
            for rec in pulled["records"] as! [[String: Any]] {
                try requireVault().applyRecord(recordJson: jsonString(rec))
                baseVersions[rec["id"] as! String] = (rec["version"] as! Int)
            }
            cursor = pulled["cursor"] as! Int
            try await pushChanges()
            syncState = .synced
        } catch {
            syncState = .error
        }
    }

    private func pushChanges() async throws {
        guard let client = api else { syncState = .pending; return }
        let records = jsonArray(try requireVault().records())
        for rec in records {
            let id = rec["id"] as! String
            let base = baseVersions[id] ?? 0
            if (rec["version"] as! Int) == base { continue }
            let res = try await client.push(record: rec, baseVersion: base)
            baseVersions[id] = res["new_version"] as! Int
            cursor = max(cursor, res["cursor"] as! Int)
        }
        cache.save(cryptoJson: cache.cryptoJson ?? "{}", recordsJson: try requireVault().records())
    }

    private func rememberBaseVersions(_ records: [[String: Any]]) {
        for rec in records { baseVersions[rec["id"] as! String] = (rec["version"] as! Int) }
    }
}

// MARK: - JSON helpers

func jsonObject(_ s: String) throws -> [String: Any] {
    try JSONSerialization.jsonObject(with: Data(s.utf8)) as! [String: Any]
}
func jsonArray(_ s: String) -> [[String: Any]] {
    (try? JSONSerialization.jsonObject(with: Data(s.utf8)) as? [[String: Any]]) ?? []
}
func jsonString(_ any: Any) -> String {
    let data = try! JSONSerialization.data(withJSONObject: any)
    return String(decoding: data, as: UTF8.self)
}

extension String {
    func trimmingTrailingSlash() -> String {
        hasSuffix("/") ? String(dropLast()) : self
    }
}

File: mobile/android/app/src/main/java/au/com/rodoskosmos/vault/VaultManager.kt
package au.com.rodoskosmos.vault

import android.content.Context
import au.com.rodoskosmos.vault.data.VaultCache
import au.com.rodoskosmos.vault.net.ApiClient
import org.json.JSONArray
import org.json.JSONObject
import uniffi.vault_mobile.VaultHandle
import uniffi.vault_mobile.benchmarkKdf
import uniffi.vault_mobile.defaultKdfParams
import uniffi.vault_mobile.generatePassword
import uniffi.vault_mobile.opaqueLoginFinish
import uniffi.vault_mobile.opaqueLoginStart
import uniffi.vault_mobile.ratePasswordStrength

enum class SyncState { SYNCED, PENDING, ERROR, OFFLINE }

/**
 * Process-global vault session shared by the UI, the AutofillService, and the
 * CredentialProvider. Holds the unlocked native vault (UniFFI) in memory only.
 */
object VaultManager {
    private var vault: VaultHandle? = null
    private var api: ApiClient? = null
    private lateinit var cache: VaultCache
    private var cursor: Long = 0
    private val baseVersions = HashMap<String, Long>()
    var syncState: SyncState = SyncState.SYNCED
        private set

    fun init(context: Context) {
        if (!::cache.isInitialized) cache = VaultCache(context.applicationContext)
    }

    val isUnlocked: Boolean get() = vault != null
    val hasCache: Boolean get() = ::cache.isInitialized && cache.hasCache()

    // --- unlock paths ------------------------------------------------------

    /** Online unlock: OPAQUE login, fetch crypto + records, open the vault. */
    fun unlockOnline(instanceUrl: String, username: String, password: String, totp: String?) {
        val client = ApiClient(instanceUrl.trimEnd('/'))
        val ls = JSONObject(opaqueLoginStart(password))
        val start = client.loginStart(username, ls.getString("message"))
        val finalization = opaqueLoginFinish(ls.getString("state"), password, start.getString("credential_response"))
        val outcome = client.loginFinish(start.getString("flow_id"), finalization, "Android", totp)
        if (!outcome.has("access_token")) throw IllegalStateException("second factor required")
        client.accessToken = outcome.getString("access_token")

        val crypto = client.accountCrypto()
        val pulled = client.pull(0)
        cursor = pulled.getLong("cursor")
        val records = pulled.getJSONArray("records")
        rememberBaseVersions(records)
        vault = VaultHandle.unlock(password, crypto.toString(), records.toString())
        api = client
        cache.save(crypto.toString(), records.toString())
        syncState = SyncState.SYNCED
    }

    /** Offline unlock from the encrypted cache with the master password. */
    fun unlockOffline(password: String) {
        val crypto = cache.cryptoJson() ?: error("no cache")
        val records = cache.recordsJson() ?: "[]"
        vault = VaultHandle.unlock(password, crypto, records)
        rememberBaseVersions(JSONArray(records))
        syncState = SyncState.OFFLINE
    }

    /** Biometric unlock: the account key comes from the Keystore, crypto and
     * records from the encrypted cache. */
    fun unlockBiometric(accountKeyB64: String) {
        val crypto = cache.cryptoJson() ?: error("no cache")
        val records = cache.recordsJson() ?: "[]"
        vault = VaultHandle.unlockWithAccountKey(accountKeyB64, crypto, records)
        rememberBaseVersions(JSONArray(records))
        syncState = SyncState.OFFLINE
    }

    fun exportAccountKey(): String = requireVault().exportAccountKey()

    fun lock() {
        vault?.destroy()
        vault = null
        api = null
        baseVersions.clear()
        cursor = 0
    }

    // --- items -------------------------------------------------------------

    private fun requireVault(): VaultHandle = vault ?: error("locked")

    fun listActive(): List<String> = requireVault().listActive()
    fun search(query: String): List<String> = requireVault().search(query)
    fun candidatesFor(url: String): List<String> = requireVault().candidatesFor(url)
    fun getItem(id: String): JSONObject = JSONObject(requireVault().getItem(id))

    /** Distinct hostnames across stored login URIs (for Digital Asset Links). */
    fun itemDomains(): List<String> {
        val out = LinkedHashSet<String>()
        for (id in listActive()) {
            val uris = getItem(id).optJSONObject("data")?.optJSONArray("uris") ?: continue
            for (i in 0 until uris.length()) {
                hostOf(uris.getJSONObject(i).optString("value"))?.let { out.add(it) }
            }
        }
        return out.toList()
    }

    private fun hostOf(url: String): String? = try {
        java.net.URI(if (url.contains("://")) url else "https://$url").host
    } catch (e: Exception) { null }

    fun createLogin(title: String, username: String, password: String, uri: String): String {
        val content = JSONObject()
            .put("title", title)
            .put("data", JSONObject()
                .put("type", "login").put("username", username).put("password", password)
                .put("uris", JSONArray().put(JSONObject().put("value", uri).put("match_rule", "base_domain"))))
            .put("notes", "").put("tags", JSONArray()).put("favorite", false)
            .put("custom_fields", JSONArray())
        val id = requireVault().createItem(content.toString())
        pushChanges()
        return id
    }

    fun updateItem(id: String, content: JSONObject) {
        requireVault().updateItem(id, content.toString())
        pushChanges()
    }

    fun generatePassword(): String =
        generatePassword("""{"length":20,"lowercase":true,"uppercase":true,"digits":true,"symbols":true}""")

    fun strength(pw: String): JSONObject = JSONObject(ratePasswordStrength(pw))

    fun negotiateParams(): String = try { benchmarkKdf(500.0) } catch (e: Exception) { defaultKdfParams() }

    // --- sync --------------------------------------------------------------

    fun sync() {
        val client = api ?: run { syncState = SyncState.OFFLINE; return }
        try {
            val pulled = client.pull(cursor)
            val records = pulled.getJSONArray("records")
            for (i in 0 until records.length()) {
                val rec = records.getJSONObject(i)
                requireVault().applyRecord(rec.toString())
                baseVersions[rec.getString("id")] = rec.getLong("version")
            }
            cursor = pulled.getLong("cursor")
            pushChanges()
            cache.save(cache.cryptoJson() ?: "{}", requireVault().records())
            syncState = SyncState.SYNCED
        } catch (e: Exception) {
            syncState = SyncState.ERROR
        }
    }

    private fun pushChanges() {
        val client = api ?: run { syncState = SyncState.PENDING; return }
        val records = JSONArray(requireVault().records())
        for (i in 0 until records.length()) {
            val rec = records.getJSONObject(i)
            val id = rec.getString("id")
            val base = baseVersions[id] ?: 0
            if (rec.getLong("version") == base) continue
            try {
                val res = client.push(rec, base)
                baseVersions[id] = res.getLong("new_version")
                cursor = maxOf(cursor, res.getLong("cursor"))
            } catch (e: Exception) {
                syncState = SyncState.PENDING
            }
        }
        cache.save(cache.cryptoJson() ?: "{}", requireVault().records())
    }

    private fun rememberBaseVersions(records: JSONArray) {
        for (i in 0 until records.length()) {
            val rec = records.getJSONObject(i)
            baseVersions[rec.getString("id")] = rec.getLong("version")
        }
    }
}
